core/vcpu.c

```diff
@@ -2067,551 +1957,164 @@ static int vcpu_simple_decode(struct vcpu_t *vcpu, struct decode *dc)
-        case 0x23: {  // AND reg/mem => reg, 16/32/64-bit
-            if (modrm.mod == 3) {
-                hax_error("Unexpected reg-to-reg AND instruction\n");
-                goto case_unexpected_opcode;
-            }
+static uint64_t vcpu_get_segment_base(void *obj, uint32_t segment)
+{
+    struct vcpu_t *vcpu = obj;
+    switch (segment) {
+    case SEG_CS:
+        return vcpu->state->_cs.base;
+    case SEG_DS:
+        return vcpu->state->_ds.base;
+    case SEG_ES:
+        return vcpu->state->_es.base;
+    case SEG_FS:
+        return vcpu->state->_fs.base;
+    case SEG_GS:
+        return vcpu->state->_gs.base;
+    case SEG_SS:
+        return vcpu->state->_ss.base;
+    default:
+        return vcpu->state->_ds.base;
+    }
+}
 
-            dc->opcode_dir = OPCODE_AND_IOMEM_TO_REG;
-            dc->size = instr[num] == 0x22 ? 1 : operand_size;
-            dc->reg_index = modrm.reg + (rex_r ? 8 : 0);
-            break;
-        }
-        // TODO: Handle 0x08, 0x09, 0x80 (/1), 0x81 (/1) and 0x83 (/1)
-        case 0x0a:    // OR reg/mem => reg, 8-bit
-        case 0x0b: {  // OR reg/mem => reg, 16/32/64-bit
-            if (modrm.mod == 3) {
-                hax_error("Unexpected reg-to-reg OR instruction\n");
-                goto case_unexpected_opcode;
-            }
+static void vcpu_advance_rip(void *obj, uint64_t len)
+{
+    struct vcpu_t *vcpu = obj;
+    advance_rip(vcpu);
+}
 
-            dc->opcode_dir = OPCODE_OR_IOMEM_TO_REG;
-            dc->size = instr[num] == 0x0a ? 1 : operand_size;
-            dc->reg_index = modrm.reg + (rex_r ? 8 : 0);
-            break;
-        }
-        // TODO: Handle 0x30, 0x31, 0x80 (/6), 0x81 (/6) and 0x83 (/6)
-        case 0x32:    // XOR reg/mem => reg, 8-bit
-        case 0x33: {  // XOR reg/mem => reg, 16/32/64-bit
-            if (modrm.mod == 3) {
-                hax_error("Unexpected reg-to-reg XOR instruction\n");
-                goto case_unexpected_opcode;
-            }
+static em_status_t vcpu_read_memory(void *obj, uint64_t ea,
+                                    uint64_t *value, uint32_t size)
+{
+    struct vcpu_t *vcpu = obj;
+    uint64_t pa;
 
-            dc->opcode_dir = OPCODE_XOR_IOMEM_TO_REG;
-            dc->size = instr[num] == 0x32 ? 1 : operand_size;
-            dc->reg_index = modrm.reg + (rex_r ? 8 : 0);
-            break;
-        }
-case_unexpected_opcode:
-        default: {
-            hax_panic_vcpu(vcpu, "Unexpected MMIO instruction (opcode=0x%x,"
-                           " exit_instr_length=%u, num=%d, gpa=0x%llx,"
-                           " instr[0..5]=0x%x 0x%x 0x%x 0x%x 0x%x 0x%x)\n",
-                           instr[num], vcpu->vmx.exit_instr_length, num,
-                           dc->gpa, instr[0], instr[1], instr[2], instr[3],
-                           instr[4], instr[5]);
-            dump_vmcs(vcpu);
-            return -1;
-        }
+    vcpu_translate(vcpu, ea, 0, &pa, NULL, false);
```

Comment:
As I said, in most cases we should be able to avoid this expensive call. I'd like to hear what you think of my previous proposal, i.e. to obtain `pa` directly from hardware:

```
pa = vmread(vcpu, VM_EXIT_INFO_GUEST_PHYSICAL_ADDRESS);
```

Of course, we should first identify the opcodes that allow more than one memory operand. E.g., we could define a new flag for `em_opcode_t::flags`, or add a new field to `em_opcode_t` that indicates the maximum number of memory operands allowed.

If we are able to fix all the remaining test failures soon, we can leave this optimization for a future PR. But I think doing it now can help debug the test failures.

Reply:
Fixed (as of e6ff1d3). I've added a `uint32_t flags` argument on memory callbacks (`read_memory` and `write_memory`) to allow such optimizations. 

After checking how other hypervisors handle this, I think in the future we could drop this `flags` arguments and export several emulator "helper" functions (e.g. like KVM's `emulator_can_use_gpa(ctxt)`), so that it's the callback itself the one who checks whether the optimization is allowed, and not the emulator who preemptively checks for allowed optimizations.